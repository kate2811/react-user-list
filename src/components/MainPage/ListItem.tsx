import React, { useCallback } from 'react'
import { useRemoveUser, useUserById } from '../../modules/core/hooks'
import { Link } from 'react-router-dom'
import cx from 'classnames'

type Props = {
  id: string
}

const ListItem: React.FC<Props> = ({ id }) => {
  const removeUser = useRemoveUser()
  const user = useUserById(id)

  if (!user) {
    throw new Error('user is not found')
  }

  const { name, surname, email, age } = user
  const onRemove = useCallback(() => {
    removeUser(id)
  }, [id, removeUser])
  return (
    <tr>
      <td>{name}</td>
      <td>{surname}</td>
      <td>{email}</td>
      <td>{age}</td>
      <td>
        <button className={cx('btn', 'btn-sm', 'btn-light', 'mr-2')} onClick={onRemove}>
          Remove
        </button>
        <Link className={cx('btn', 'btn-sm', 'btn-light')} to={`/edit-user/${id}`}>
          Edit
        </Link>
      </td>
    </tr>
  )
}

export default ListItem
