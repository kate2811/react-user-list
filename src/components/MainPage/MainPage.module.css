.spinner {
  width: 40px;
  margin-top: 70px;
}

.searchForm {
  margin-bottom: 30px;
}

.tableHead {
  position: relative;
  text-transform: capitalize;
}

.tableHead__title {
  margin-right: 10px;
}

.tableHead i {
  position: absolute;
  top: 15px;
}

.tableHead i:first-of-type {
  opacity: 0.4;
}
