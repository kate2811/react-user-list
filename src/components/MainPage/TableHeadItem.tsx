import React from 'react'
import { sortParams } from '../../modules/core/types'
import style from './MainPage.module.css'
import cx from 'classnames'

type Props = {
  name: string
  isSortable: boolean
  setSortParams: (newSortField: string) => void
  sortParams: sortParams
}

const TableHeadItem: React.FC<Props> = ({ name, isSortable, setSortParams, sortParams }) => {
  return (
    <>
      {isSortable ? (
        <th scope="col" onClick={() => setSortParams(name)} className={style.tableHead}>
          <span className={style.tableHead__title}>{name}</span>
          <i className="fas fa-sort" />
          {sortParams.field === name && (
            <i className={cx('fas', sortParams.direction === 'asc' ? 'fa-sort-down' : 'fa-sort-up')} />
          )}
        </th>
      ) : (
        <th scope="col" className={style.tableHead}>
          {name}
        </th>
      )}
    </>
  )
}

export default TableHeadItem
