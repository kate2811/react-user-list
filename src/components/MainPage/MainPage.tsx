import React from 'react'
import { Link } from 'react-router-dom'
import ListItem from './ListItem'
import { useIsLoading, useUserId } from '../../modules/core/hooks'
import PageLayout from '../PageLayout'
import Loader from 'react-loader-spinner'
import cx from 'classnames'

const MainPage: React.FC = () => {
  const userIdList = useUserId()
  const isLoading = useIsLoading()

  return (
    <PageLayout title={'User list'}>
      {isLoading ? (
        <Loader type="BallTriangle" color="#000000" height={20} width={20} />
      ) : (
        <>
          <table className={cx('table', 'table-hover')}>
            <thead className="bg-light">
            <tr>
              <th scope="col">Name</th>
              <th scope="col">Surname</th>
              <th scope="col">Email</th>
              <th scope="col">Age</th>
              <th scope="col">Actions</th>
            </tr>
            </thead>
            <tbody>
            {userIdList.map((item, index) => (
              <ListItem key={index} id={item} />
            ))}
            </tbody>
          </table>
          <Link className={cx('btn', 'btn-primary', 'w-25')} to={'/add-user'}>Add user</Link>
        </>
      )}
    </PageLayout>
  )
}

export default MainPage
