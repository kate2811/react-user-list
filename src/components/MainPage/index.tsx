import React from 'react'
import MainPage from './MainPage'
import { useIsLoading, useUserId } from '../../modules/core/hooks'

export default function() {
  const userIdList = useUserId()
  const isLoading = useIsLoading()
  return <MainPage userIdList={userIdList} isLoading={isLoading} />
}
