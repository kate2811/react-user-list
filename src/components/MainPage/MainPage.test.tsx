import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import MainPage from './MainPage'
import { MemoryRouter } from 'react-router-dom'

const onClick = jest.fn()

jest.mock('./TableItem', () => ({
  __esModule: true,
  default: function TableItem(props: any) {
    return (
      <tr data-testid="tr">
        <td>{JSON.stringify(props)}</td>
      </tr>
    )
  }
}))

describe('MainPage', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('is loading', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <MainPage isLoading={true} userIdList={null} />
      </MemoryRouter>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('is loaded', () => {
    const { getAllByTestId } = render(
      <MemoryRouter>
        <MainPage isLoading={false} userIdList={['1', '2', '3']} />
      </MemoryRouter>
    )
    expect(getAllByTestId('tr').length).toEqual(3)
  })

  it('button is clickable', () => {
    const { getByTestId } = render(
      <MemoryRouter>
        <MainPage isLoading={false} userIdList={null} />
      </MemoryRouter>
    )
    fireEvent.click(getByTestId('button'))
    expect(onClick).toHaveBeenCalledTimes(1)
  })

  it('button is clickable 2', () => {
    const { getByTestId } = render(
      <MemoryRouter>
        <MainPage isLoading={false} userIdList={null} />
      </MemoryRouter>
    )
    fireEvent.click(getByTestId('button'))
    expect(onClick).toHaveBeenCalledTimes(1)
  })
})
