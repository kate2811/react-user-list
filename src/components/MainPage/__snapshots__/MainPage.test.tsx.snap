// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage is loading 1`] = `
<DocumentFragment>
  <div>
    <nav
      class="navbar navbar-dark bg-primary justify-content-start"
    >
      <span
        class="navbar-brand mb-0 h1"
      >
        UserList
      </span>
      <div
        class="navbar-nav"
      >
        <a
          class="nav-item nav-link active"
          href="/"
        >
          Home
        </a>
      </div>
    </nav>
    <nav
      aria-label="breadcrumb"
    >
      <ol
        class="breadcrumb"
      >
        <li
          class="breadcrumb-item"
        >
          <a
            href="/"
          >
            Home
          </a>
        </li>
      </ol>
    </nav>
    <div
      class="container mt-4"
    >
      <h1
        class="mb-3"
      >
        User list
      </h1>
      <div
        class="spinner mx-auto"
      >
        <div
          aria-busy="true"
          class=""
        >
          <svg
            aria-label="audio-loading"
            fill="#157FFB"
            height="40"
            viewBox="0 0 135 140"
            width="40"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="120"
              rx="6"
              width="15"
              y="10"
            >
              <animate
                attributeName="height"
                begin="0.5s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="120;110;100;90;80;70;60;50;40;140;120"
              />
              <animate
                attributeName="y"
                begin="0.5s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="10;15;20;25;30;35;40;45;50;0;10"
              />
            </rect>
            <rect
              height="120"
              rx="6"
              width="15"
              x="30"
              y="10"
            >
              <animate
                attributeName="height"
                begin="0.25s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="120;110;100;90;80;70;60;50;40;140;120"
              />
              <animate
                attributeName="y"
                begin="0.25s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="10;15;20;25;30;35;40;45;50;0;10"
              />
            </rect>
            <rect
              height="140"
              rx="6"
              width="15"
              x="60"
            >
              <animate
                attributeName="height"
                begin="0s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="120;110;100;90;80;70;60;50;40;140;120"
              />
              <animate
                attributeName="y"
                begin="0s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="10;15;20;25;30;35;40;45;50;0;10"
              />
            </rect>
            <rect
              height="120"
              rx="6"
              width="15"
              x="90"
              y="10"
            >
              <animate
                attributeName="height"
                begin="0.25s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="120;110;100;90;80;70;60;50;40;140;120"
              />
              <animate
                attributeName="y"
                begin="0.25s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="10;15;20;25;30;35;40;45;50;0;10"
              />
            </rect>
            <rect
              height="120"
              rx="6"
              width="15"
              x="120"
              y="10"
            >
              <animate
                attributeName="height"
                begin="0.5s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="120;110;100;90;80;70;60;50;40;140;120"
              />
              <animate
                attributeName="y"
                begin="0.5s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="10;15;20;25;30;35;40;45;50;0;10"
              />
            </rect>
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
